crates/astria-conductor/src/sequencer.rs

```diff
@@ -0,0 +1,365 @@
+                    warn!(
+                        error = &e as &dyn StdError,
+                        "latest sequencer block couldn't be inserted into block cache; is the \
+                         start sync height in the future?",
+                    );
+                }
+                height
+            }
+        };
+
+        let mut blocks_from_height =
+            BlocksFromHeightStream::new(start_height, latest_height, pool.clone(), 20);
+
+        let mut resubscribing = future::Fuse::terminated();
+        'reader_loop: loop {
+            select! {
+                shutdown = &mut shutdown => {
+                    let ret = if let Err(e) = shutdown {
+                        warn!(
+                            error = &e as &dyn StdError,
+                            "shutdown channel closed unexpectedly; shutting down",
+                        );
+                        Err(e).wrap_err("shut down channel closed unexpectedly")
+                    } else {
+                        info!("received shutdown signal; shutting down");
+                        Ok(())
+                    };
+                    break 'reader_loop ret;
+                }
+
+                Some(block) = blocks_from_height.next() => {
+                    if let Err(e) = sequential_blocks.insert(block) {
+                        // XXX: we could temporarily kill the subscription if we put an upper limit on the cache size
+                        warn!(error = &e as &dyn std::error::Error, "failed pushing block into cache, dropping it");
+                    }
+                }
+
+
+                Some(block) = sequential_blocks.next_block() => {
+                    if let Err(e) = executor_channel.send(block.into()) {
+                        let reason = "failed sending next sequencer block to executor";
+                        error!(
+                            error = &e as &dyn std::error::Error,
+                            reason,
+                            "exiting",
+                        );
+                        break 'reader_loop Err(e).wrap_err(reason);
+                    }
+                }
+
+                // Blocks from the sequencer subscription. Resubscribes if `None`
+                new_block = subscription.next(), if !subscription.is_done() => {
+                    match new_block {
+                        Some(Ok(block)) => {
+                            let height = block.height();
+                            if let Err(e) = sequential_blocks.insert(block) {
+                                // XXX: we could temporarily kill the subscription if we put an upper limit on the cache size
+                                warn!(error = &e as &dyn std::error::Error, "failed pushing block into cache, dropping it");
+                            } else {
+                                blocks_from_height.advance_height(height);
+                            }
```

Comment:
why not have the `BlocksFromHeightStream` advance the height internally?

Reply:
Because that information is not available within that stream. Fundamentally, `BlocksFromHeightStream` is modelled around a `FuturesUnordered` into which new requests are pushed. The assumption is that if a height is requested, then it is expected to exist. The stream will then eagerly request as many heights as possible.

On the other hand, if one requests a cometbft height that does not exist (yet), cometbft will respond with a 500. So the stream could just autoadvance its heights (or use an infinite loop), but then you'd get a ton of 500s which you need to spin on until you get an actual response.

I thought it's a cleaner solution to just set the heights externally (kind of like pushing new heights into the stream).

Note that this is changed in https://github.com/astriaorg/astria/pull/725